_posts/2024-11-19-datafusion-python-udf-comparisons.md

```diff
@@ -0,0 +1,623 @@
+```
+
+Since we are doing a `sum` we can keep a single value as our internal state. When we call `update()`
+we will process a single array and update the internal state, which we share with the `state()`
+function. For larger batches we may `merge()` these states. It is important to note that the
+`states` in the `merge()` function are an array of the values returned from `state()`. It is
+entirely possible that the `merge` function is significantly different than the `update`, though in
+our example they are very similar.
+
+## User Defined Window Functions
+
+Writing a user defined window function is slighlty more complex than an aggregate function due
+to the variety of ways that window functions are called. I recommend reviewing the
+[online documentation](https://datafusion.apache.org/python/user-guide/common-operations/udf-and-udfa.html)
+for a description of which functions need to be implemented. The details of how to implement
+these generally follow the same patterns as described above for aggregate functions.
+
+## Performance Comparison
+
+For the scalar functions above, we performed a timing evaluation, repeating the operation 100
+times. For this simple example these are our results.
+
+```
++-----------------------------+--------------+---------+
+| approach                    | Average Time | Std Dev |
++-----------------------------+--------------+---------+
+| python udf                  | 4.969        | 0.062   |
+| simple filter               | 1.075        | 0.022   |
+| explicit filter             | 0.685        | 0.063   |
+| pyarrow compute             | 0.529        | 0.017   |
+| arrow rust compute          | 0.511        | 0.034   |
+| arrow rust compute as class | 0.502        | 0.011   |
+| rust custom iterator        | 0.478        | 0.009   |
++-----------------------------+--------------+---------+
+```
+
+As expected, the conversion to Python objects is by far the worst performance. As soon as we drop
+into using any functions that keep the data entirely on the Rust side we see a near 10x speed
+improvement. Then as we increase our complexity from using PyArrow compute functions to
+implementing the UDF in Rust we see incremental improvements. Our fastest approach - iterating
+through the arrays ourselves does operate nearly 10% faster than the PyArrow compute approach.
+
+## Final Thoughts and Recommendations
+
+For anyone who is curious about [DataFusion](https://datafusion.apache.org/) I highly recommend
+giving it a try. This post was designed to make it easier for new users to the Python implementation
+to work with User Defined Functions by giving a few examples of how one might implement these.
+
+When it comes to designing UDFs, I strongly recommend seeing if you can write your UDF using
+[PyArrow functions](https://arrow.apache.org/docs/python/api/compute.html) rather than pure Python
+objects. These will give you enormous speed benefits. If you must do something that isn't well
```

Comment:
Some of users jump directly into conclusions sections, so I'd add an short notice, what was speed benefits